influxdata/tokenizer.go

```diff
@@ -7,91 +7,187 @@ import (
 	buf []byte
 
 	// complete holds whether the data in buffer
 	// is known to be all the data that's available.
 	complete bool
 
-	// r holds the read position in buf.
-	r int
+	// skipping holds whether we will need
+	// to return the values that we're tokenizing.
+	skipping bool
+
+	// r0 holds the earliest read position in buf.
+	// Data in buf[0:r0] is considered to be discarded.
+	r0 int
+
+	// r1 holds the read position in buf. Data in buf[r1:] is
+	// next to be read. Data in buf[len(buf):cap(buf)] is
+	// available for reading into.
+	r1 int
+
+	// escBuf holds a buffer for unescaped characters.
+	escBuf []byte
 
 	// err holds any non-EOF error that was returned from rd.
 	err error
 }
 
-// take returns the next slice of bytes within t.buf that are in the given set
-// starting at t.r + from, reading more data as needed.
-//
-// It does not update t.r.
-func (t *Tokenizer) take(set *byteSet, from int) []byte {
-	n := from
+// NewTokenizer returns a tokenizer that splits the line-protocol text
+// inside buf. The text in buf represent a complete line (or multiple lines) - that
+// is, the tokenization will not work correctly if presented with a line
+// split at an arbitrary place.
+func NewTokenizerWithBytes(buf []byte) *Tokenizer {
+	return &Tokenizer{
+		buf:      buf,
+		complete: true,
+		escBuf:   make([]byte, 0, 512),
+	}
+}
+
+func NewTokenizer(r io.Reader) *Tokenizer {
+	return &Tokenizer{
+		rd: r,
+	}
+}
+
+// take returns the next slice of bytes that are in the given set
+// reading more data as needed. It updates t.r1.
+func (t *Tokenizer) take(set *byteSet) []byte {
+	// Note: use a relative index for start because absolute
+	// indexes aren't stable (the contents of the buffer can be
+	// moved when reading more data).
+	start := t.r1 - t.r0
 outer:
 	for {
-		if !t.ensure(n + 1) {
+		if !t.ensure(1) {
 			break
 		}
-		buf := t.buf[t.r+n:]
+		buf := t.buf[t.r1:]
 		for i, c := range buf {
 			if !set.get(c) {
-				n += i
+				t.r1 += i
 				break outer
 			}
 		}
-		n += len(buf)
+		t.r1 += len(buf)
 	}
-	return t.buf[t.r+from : t.r+n]
+	return t.buf[t.r0+start : t.r1]
+}
+
+// takeEsc is like take except that escaped characters also count as
+// part of the set. The escapeTable determines which characters
+// can be escaped.
+// It returns the unescaped string (unless t.skipping is true, in which
+// case it doesn't need to go to the trouble of unescaping it).
+func (t *Tokenizer) takeEsc(set *byteSet, escapeTable *[256]byte) []byte {
+	// start holds the offset from r0 of the start of the taken slice.
+	// Note that we can't use t.r1 directly, because the offsets can change
+	// when the buffer is grown.
+	start := t.r1 - t.r0
+
+	// startUnesc holds the offset from t0 of the start of the most recent
+	// unescaped segment.
+	startUnesc := start
+
+	// startEsc holds the index into r.escBuf of the start of the escape buffer.
+	startEsc := len(t.escBuf)
+outer:
+	for {
+		if !t.ensure(1) {
```

Comment:
note to me: `ensure(1)` does not read in just one more byte. If there are many bytes to read, it will read min(bytes_left, some_max_min_value).

Reply:
it actually doesn't read any bytes at all unless it has to. it ensures that there's at least one byte available in the "unread" part of the buffer. only if the buffer is empty will it call Read.

Reply:
> it actually doesn't read any bytes at all unless it has to.

yes yes - there was an elided `if this needs to read bytes`, it does not read just one (unless there's actually just one more byte to read)